File: problems/divide-and-conquer/dc_problems_test.go
package dc

import (
	"math"
	"reflect"
	"testing"
)

func TestMaxSubarray(t *testing.T) {
	nums := []int{-2, 1, -3, 4, -1, 2, 1, -5, 4}
	expected := 6
	got := MaxSubarray(nums)
	if got != expected {
		t.Errorf("MaxSubarray(%v) = %d; want %d", nums, got, expected)
	}
}

func TestStrassenMultiply2x2(t *testing.T) {
	A := [][]int{{1, 2}, {3, 4}}
	B := [][]int{{5, 6}, {7, 8}}
	expected := [][]int{{19, 22}, {43, 50}}
	got := StrassenMultiply2x2(A, B)
	if !reflect.DeepEqual(got, expected) {
		t.Errorf("StrassenMultiply2x2() = %v; want %v", got, expected)
	}
}

func TestQuickSort(t *testing.T) {
	nums := []int{9, 7, 5, 11, 12, 2, 14, 3, 10, 6}
	expected := []int{2, 3, 5, 6, 7, 9, 10, 11, 12, 14}
	QuickSort(nums)
	if !reflect.DeepEqual(nums, expected) {
		t.Errorf("QuickSort() = %v; want %v", nums, expected)
	}
}

func TestRandomizedSelect(t *testing.T) {
	nums := []int{9, 7, 5, 11, 12, 2, 14, 3, 10, 6}
	// Sorted: 2, 3, 5, 6, 7, 9, 10, 11, 12, 14
	kthSmallest := RandomizedSelect(nums, 0, len(nums)-1, 5) // 5th smallest is 7
	if kthSmallest != 7 {
		t.Errorf("RandomizedSelect(k=5) = %d; want 7", kthSmallest)
	}
}

func TestClosestPair(t *testing.T) {
	points := []Point{
		{X: 2, Y: 3},
		{X: 12, Y: 30},
		{X: 40, Y: 50},
		{X: 5, Y: 1},
		{X: 12, Y: 10},
		{X: 3, Y: 4},
	}
	expected := math.Sqrt(2) // between (2,3) and (3,4) is ~1.4142
	got := ClosestPair(points)
	if math.Abs(got-expected) > 1e-4 {
		t.Errorf("ClosestPair() = %f; want %f", got, expected)
	}
}

File: problems/divide-and-conquer/DCProblemsTest.java
import java.util.Arrays;

public class DCProblemsTest {

    public static void main(String[] args) {
        testMaxSubarray();
        testStrassenMultiply2x2();
        testQuickSort();
        testRandomizedSelect();
        testClosestPair();
        System.out.println("ALL DIVIDE AND CONQUER TESTS PASSED!");
    }

    private static void assertTrue(boolean condition, String message) {
        if (!condition) {
            throw new AssertionError("Test Failed: " + message);
        }
    }

    private static void testMaxSubarray() {
        int[] nums = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
        assertTrue(DCProblems.maxSubArray(nums) == 6, "maxSubArray");
    }

    private static void testStrassenMultiply2x2() {
        int[][] A = {{1, 2}, {3, 4}};
        int[][] B = {{5, 6}, {7, 8}};
        int[][] C = DCProblems.strassenMultiply2x2(A, B);
        assertTrue(C[0][0] == 19 && C[0][1] == 22 && C[1][0] == 43 && C[1][1] == 50, "strassen2x2");
    }

    private static void testQuickSort() {
        int[] nums = {9, 7, 5, 11, 12, 2, 14, 3, 10, 6};
        int[] expected = {2, 3, 5, 6, 7, 9, 10, 11, 12, 14};
        DCProblems.quickSort(nums, 0, nums.length - 1);
        assertTrue(Arrays.equals(nums, expected), "quickSort");
    }

    private static void testRandomizedSelect() {
        int[] nums = {9, 7, 5, 11, 12, 2, 14, 3, 10, 6};
        int kth = DCProblems.randomizedSelect(nums, 0, nums.length - 1, 5);
        assertTrue(kth == 7, "randomizedSelect");
    }

    private static void testClosestPair() {
        DCProblems.Point[] pts = {
            new DCProblems.Point(2, 3),
            new DCProblems.Point(12, 30),
            new DCProblems.Point(40, 50),
            new DCProblems.Point(5, 1),
            new DCProblems.Point(12, 10),
            new DCProblems.Point(3, 4)
        };
        double d = DCProblems.closestPair(pts);
        assertTrue(Math.abs(d - Math.sqrt(2)) < 1e-4, "closestPair");
    }
}
